File: 2501/2501.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    2501.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2023/09/29 23:21:51 by jihoolee          #+#    #+#              #
#    Updated: 2023/09/30 00:41:47 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #


def get_factors(num: int) -> list[int]:
    factors: list[int] = [1, num]

    i = 2
    while (i ** 2 <= num):
        if (num % i == 0):
            factors.insert(len(factors) // 2, i)
            if (i ** 2 != num):
                factors.insert(len(factors) // 2 + 1, num // i)
        i += 1
    return factors


def main() -> None:
    num, k = map(int, input().split(' '))
    try:
        print(get_factors(num)[k - 1])
    except IndexError as e:
        print(0)


if (__name__ == '__main__'):
    main()

File: 2501/2501.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   2501.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/09/29 15:17:37 by jihoolee          #+#    #+#             */
/*   Updated: 2023/09/29 23:20:56 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <vector>

std::vector<int> get_factors(int num) {
  std::vector<int> factors = {1, num};

  for (int i = 2; i * i <= num; ++i) {
    if (num % i == 0) {
      factors.insert(factors.begin() + (factors.size() / 2), i);
      if (i * i != num)
        factors.insert(factors.begin() + (factors.size() / 2 + 1), num / i);
    }
  }
  return factors;
}

int main(void) {
  int               num;
  int               k;
  std::vector<int>  factors;

  std::cin >> num >> k;
  get_factors(num).swap(factors);
  try {
    std::cout << factors.at(k - 1);
  } catch (const std::out_of_range& oor) {
    std::cout << 0;
  }
  return 0;
}
